go-controller/pkg/libovsdbops/acl.go

```diff
@@ -137,7 +142,7 @@ func CreateOrUpdateACLsOps(nbClient libovsdbclient.Client, ops []libovsdb.Operat
 		opModel := operationModel{
 			Model:          acl,
 			ModelPredicate: func(item *nbdb.ACL) bool { return isEquivalentACL(item, acl) },
-			OnModelUpdates: onModelUpdatesAllNonDefault(),
```

Comment:
@npinaeva can you explain a little more how this was not updating acl severity to nil? onModelUpdatesAllNonDefault() would return an empty slice. Then that would get passed down into libovsdb signalling the default to update all fields to their default value. Why is this not working correctly? This looks like a string pointer field, and iirc we had some issues in libovsdb with those before. I think it eventually makes its way to ```ovsElem, err := ovsdb.NativeToOvs(column, nativeElem)``` where it should be considered and ovs set.

Reply:
IFAIK, when the default value is passed to update, it is ignored
```
	// Update returns the operations needed to update any number of rows according
	// to the data in the given model.
	// By default, all the non-default values contained in model will be updated.
	// Optional fields can be passed (pointer to fields in the model) to select the
	// the fields to be updated
	Update(model.Model, ...interface{}) ([]ovsdb.Operation, error)
```
it comes to this
```
// NewRow transforms an orm struct to a map[string] interface{} that can be used as libovsdb.Row
// By default, default or null values are skipped. This behavior can be modified by specifying
// a list of fields (pointers to fields in the struct) to be added to the row
func (m Mapper) NewRow(data *Info, fields ...interface{}) (ovsdb.Row, error) {
```
so when you pass empty slice to update, it is the "default mode" which skips default values, and if you pass the fields you want to update directly it will include them even if the value is default.
there are some examples for other object types like https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/libovsdbops/address_set.go#L22

Reply:
it also links to https://github.com/ovn-org/libovsdb/issues/226

Reply:
The default behavior of libovsdb is to only update non-nil non-zero non-empty values.

The `getNonZero<type>MutableFields` was used to interpret empty slices/maps as a magic value to indicate to libovsdb that we would like them cleared out. The use of this magic value was arguably not great but came out as natural with the way we processed LBs and was used only there.

This was done before pointers were introduced for optional attributes. Unfortunately we don't have a magic value for pointers to use in the same way.

So we don't have a consistent way to clear out values. Additionally to this and the default behavior, on some other places we assume statically the attributes that should be updated. On other places we just pass those in. 

Here, `getNonZeroACLMutableFields` I don't think it's doing what the method name implies it's doing. Otherwise, the implementation would be similar to `getNonZeroLoadBalancerMutableFields`.

I think we this should be akin to a `onModelUpdatesAll` just as we have a `onModelUpdatesAllNonDefault` and `onModelUpdatesNone`.

To have some level of consistency we could have a variation of the `CreateOrUpdate` methods where the current ones will just have a default behaviour and the variation which perhaps could be called `CreateOrUpdate...WithMode` would let the user pass one of `onModelUpdatesAll`, `onModelUpdatesAllNonDefault`, `onModelUpdatesNone` or a custom function that will let it choose exactly what to update

Reply:
I also don't think the method name is correct, but not sure if I follow what you are saying @jcaamano. The words "non zero" seems wrong to me. These fields may be zero, nil, or have a value and they will be updated no matter what by providing the field. I think simply saying getACLNonMutableFields makes more sense. The consequence to do doing this is now all of these fields will be updated anytime someone calls CreateOrUpdateACLsOps. I hope we don't break any other places of code where people depended on this behavior.

That being said making this even more generic to an "onModelUpdatesAll" where we get all fields generically from any type seems logical, but that opens up the risk even more that someone could replace a CreateOrUpdate* method with this and we hit some regression.

It seems like changing libovsdb to treat a nil pointer as non default is not an option, because the default value given as an api is a nil pointer. I guess one possibility is passing pointer to an empty string, and considering that non-default. Then changing our build ACL code to always pass an empty string pointer when severity is not specified. wdyt?

Reply:
> Your option works for me as well though. For consistency, we will need similar build methods for other types as the one we have for ACLs.

Additionally, the libovsdbops user may still want to update other columns or may want to clear out ACLSeverity. Maybe in the case of ACLs we always update them fully in one go so this does not apply. But when talking about consistency, I was thinking about what to do in the general case and in the general case, the libovsdbops user may want to partially update a row or to clear out some row column. So it needs to be made aware of these magic values or we need to have different build methods for each scenario of each specific type.

Reply:
Just to clarify my opinion for this PR. I think what Nadia did is OK, but I would like the name of the method changed. For solving this problem generically for all types we can defer to later.

Reply:
We may need a separate PR for that, for now all the methods that have specified updatable fields are called `CreateOrUpdate`, so I don't think it makes sense to change that for acls only, there is even `getAllUpdatableFields` function, that is not even used by all objects that need to specify updatable fields. Anyway, for now I ma changing the method name to `getACLMutableFields` only, lmk if there is anything else that you think would be good to update in this PR

Reply:
@jcaamano you OK if we defer that to another PR?